### ISSUE TITLE
Review nxcomp to drop all compatibility code for versions prior to 3.5.x

### ISSUE THREAD

**Issue by fcarvajaldev** [ArcticaProject/nx-libs#108]
For clean-up purposes all compatibility code for nxcomp < 3.5.x should be dropped. Anyone who is still using nxcomp 3.4 and earlier, should really stop using that code and upgrade.

At this moment nxcomp has compatibility code for versions up to 1.5.0 (protocol step 6) and there are many blocks of conditional code depending on the protocol step value.

Setting a new compatibility limit at 3.5.x versions would imply working only with protocol step 10 (which also includes previous ones), so the conditional code depending on the protocol step value could be simplified.

Those blocks which are specific for an older protocol step can be removed and every block that conforms to protocol step values previous to the current value 10 and forward will remain in the code but now without any condition, as it is assumed that if protocol's version negotiation doesn't fail and exits, then both sides (local and remote) will be working with protocol step 10 (or with a higher one in the future).

**Comment by sunweaver** [comment]
Thanks for looking into this.

One question: would it make sense to keep the step 10 condition checking (even if always resulting as true) as documentation of history? If the code becomes much more readable when dropping the if clauses, forget my question.

Another reason for keeping the conditionals could be future work on the protocol. If we know where we checked for conditions earlier, we maybe are able to more easily identify where new condition checks have to be put later on. Just theoretical thinking... without having looked at the specific code blocks, actually.

**Comment by fcarvajaldev** [comment]
Both reasons seem reasonable but I see some drawbacks in keeping the conditionals.

The main problem is that keeping the conditionals (even if always resulting as true) would also require to keep the corresponding else branches just as they are right now, or at least with an error handling procedure to manage the case of working with an older protocol version (despite this should not really happen at all).

As I see it this would leave the code much more the same as it is right now and the clean-up goal wouldn't be accomplished. But not including those else branches would lead to inconsistent and incomplete code, possibly leading to confusion and errors when trying future work on the protocol.

So in my opinion the code becomes more readable when dropping the conditional clauses, so we can get rid of blocks, methods and classes which are no longer used. In such cases to provide some documentation of history my first intention is to leave a comment instead of the conditional, dropping the whole else branch which no longer makes sense.

For example, in ClientChannel::handleRead the conditional from line 1290:

```
          if (control -> isProtoStep7() == 1)
          {
            encodeBuffer.encodeFreeXidValue(GetULONG(inputMessage + 4, bigEndian_),
                               clientCache_ -> freeDrawableCache);
          }
          else
          {
            unsigned int pixmap = GetULONG(inputMessage + 4, bigEndian_);
            unsigned int diff = pixmap - clientCache_ -> createPixmapLastId;
            if (diff == 0)
            {
              encodeBuffer.encodeBoolValue(1);
            }
            else
            {
              encodeBuffer.encodeBoolValue(0);
              clientCache_ -> createPixmapLastId = pixmap;
              encodeBuffer.encodeValue(diff, 29, 4);
            }
          }
```

, would change into something like this now that we use protocol step 10 and it conforms with protocol step 7 condition:

```
          //Since ProtoStep7
          encodeBuffer.encodeFreeXidValue(GetULONG(inputMessage + 4, bigEndian_),
                             clientCache_ -> freeDrawableCache);
```

With this simple kind of comments there should be easier to keep some track of protocol version changes. Of course, any further need of revision for future work on the protocol would require the use of git to search for the differences, but only if there is a change somehow related with the old conditionals.

**Comment by sunweaver** [comment]
Ack. Please go ahead with what you proposed. The above code example look fine. Than you!!!

**Comment by uli42** [comment]
On Thu, Mar 31, 2016 at 1:33 PM, fcarvajaldev notifications@github.com wrote:

> For example, in ClientChannel::handleRead the conditional from line 1290:
> 
> ```
>       if (control -> isProtoStep7() == 1)
>       {
>         encodeBuffer.encodeFreeXidValue(GetULONG(inputMessage + 4, bigEndian_),
>                            clientCache_ -> freeDrawableCache);
>       }
>       else
>       {
>         unsigned int pixmap = GetULONG(inputMessage + 4, bigEndian_);
>         unsigned int diff = pixmap - clientCache_ -> createPixmapLastId;
>         if (diff == 0)
>         {
>           encodeBuffer.encodeBoolValue(1);
>         }
>         else
>         {
>           encodeBuffer.encodeBoolValue(0);
>           clientCache_ -> createPixmapLastId = pixmap;
>           encodeBuffer.encodeValue(diff, 29, 4);
>         }
>       }
> ```
> 
> , would change into something like this now that we use protocol step 10 and it conforms with protocol step 7 condition:
> 
> ```
>       //Since ProtoStep7
>       encodeBuffer.encodeFreeXidValue(GetULONG(inputMessage + 4, bigEndian_),
>                          clientCache_ -> freeDrawableCache);
> ```

How about putting the whole code block into a function/method and just
call it here? This way you could have both: readable code and
compatibility.

Since development targets Linux in the first place it might become (or
already is?) difficult to compile for old platforms. So if you have
some legacy systems that cannot be installed with a current
implementation of NX you will lose the ability to contact them.

Uli

**Comment by salva** [comment]
@uli42: our aim is to simplify and improve the nxcomp code in order to get it to a state where we could be comfortable adding new features.

As a first step we are looking for code branches that depend on old protocol versions (that we assume nobody really uses anymore) and removing them. But then, we have found that there is going to be a cascade effect that would allow us to remove even more dead code (i.e. classes, and functions becoming unused and code that depends on flags that now are effectively fixed to some value). This is low-hanging fruit.

If we just move the code to different functions or locations, none of that would happen. We would still continue to have as much code as before. There is no gain there.

**Comment by nitomartinez** [comment]
My 2 cents:
- Remove it
- Put a comment and referencing this issue, so we can find what it was before if needed

Second alternative
- Create a README.protocol-compatibility.txt with the same info

**Comment by sunweaver** [comment]
@uli42: I am about to merge the PR referenced above (https://github.com/ArcticaProject/nx-libs/pull/111). I understand your concerns regarding old platforms.

Do we agree that keeping 3.5.x proto compatibility for as long as possible / sensible will suffice your needs? What we are planning to drop here is proto support for nxcomp based builds << 3.5.0.

Greets,
Mike

**Comment by uli42** [comment]
On Wed, Apr 13, 2016 at 10:28 AM, Mike Gabriel notifications@github.com wrote:

> @uli42: I am about to merge the PR referenced above (#111). I understand your concerns regarding old platforms.
> 
> Do we agree that keeping 3.5.x proto compatibility for as long as possible / sensible will suffice your needs? What we are planning to drop here is proto support for nxcomp based builds << 3.5.0.

It is not really a need. But generally I do not like throwing out
stuff that works. Might be useful one day.

But I am in no way blocking. Do whatever is best for the project. I
just wanted to speak out my concerns.

Uli

**Comment by sunweaver** [comment]
Closed by 03b65211260ff757b35a367f8936fd5e882ce56c.
